rust/kernel/error.rs

```diff
@@ -58,6 +58,26 @@ impl Error {
     pub fn to_kernel_errno(&self) -> c_types::c_int {
         self.0
     }
+
+    /// Creates an [`Error`] from a kernel pointer.
+    ///
+    /// # Pointer value range
+    ///
+    /// According to `include/linux/err.h`:
+    /// - [0, .., `core::usize::MAX - bindings::MAX_ERRNO`) is the range for
+    ///   normal values of pointer.
+    ///
+    /// - [`core::usize::MAX - bindings::MAX_ERRNO`,..,`core::usize::MAX] is the
+    ///   range for error values stored in pointer.
+    pub fn from_kernel_ptr(ptr: *mut c_types::c_void) -> Option<Error> {
```

Comment:
The kernel pointer we want to convert *from* isn't necessarily a `*mut c_void`, right? It's always a pointer to a `bindings::` type, or `c_void`. Casts are not idiomatic, so can we avoid the cast by using a template here? E.g. `from_kernel_ptr<T>(ptr: *mut T)`.

Also, why have a `pub` function that returns `Option<Error>`? Where would this be used outside of `error.rs`? Would it be simpler to have a single `pub ptr_to_result()` function that returns `Result<T>`?

Reply:
> The kernel pointer we want to convert _from_ isn't necessarily a `*mut c_void`, right? It's always a pointer to a `bindings::` type, or `c_void`. Casts are not idiomatic, so can we avoid the cast by using a template here? E.g. `from_kernel_ptr<T>(ptr: *mut T)`.
> 

Try to avoid template functions when possible. Would unnecessary template functions increase the code size? For example, if we have `from_kernel_ptr<i32>` and `from_kernel_ptr<i8>`, there will be two functions instead of one. 

> Also, why have a `pub` function that returns `Option<Error>`? Where would this be used outside of `error.rs`? Would it be simpler to have a single `pub ptr_to_result()` function that returns `Result<*mut T>`?

Oh, I just thought this might be useful, some one may use it in the future. I'm OK if we remove it.

Reply:
> Try to avoid template functions when possible. Would unnecessary template functions increase the code size? For example, if we have `from_kernel_ptr<i32>` and `from_kernel_ptr<i8>`, there will be two functions instead of one.

Perhaps the template instantiation would just get optimized away by LLVM? Maybe `from_kernel_ptr<i32>` and `from_kernel_ptr<i8>` will both result in the same function (inlined or not) which just calls `IS_ERR()` and `ERR_PTR()` ? Or perhaps one or two asm instructions if the kernel helpers are not used.

Maybe that's worth verifying? Premature optimization and roots of all evil etc :)

Reply:
I took a quick look at the asm generated for this:
```rust
    pub fn init_mmio_platform_resource(
        pdev: &mut PlatformDevice,
        index: u32,
        cfg: &RegmapConfig,
    ) -> Result<Self> {
        let iomem = Self::devm_platform_ioremap_resource(pdev, index)?;
        Self::devm_regmap_init_mmio(pdev, iomem, cfg)
    }
```
where
```rust
pub(crate) unsafe fn ptr_err_check<T>(ptr: *mut T) -> Result<*mut T> {
    if rust_helper_is_err(ptr as _) {
        return Err(Error::from_kernel_errno(
            rust_helper_ptr_err(ptr as _) as i32
        ));
    }
    Ok(ptr)
}
```
and it does look like the `ptr_err_check<T>()` template function is simply optimized away:
```asm
0001407c <_RNvMs0_NtCsbDqzXfLQacH_6kernel6regmapNtB5_6Regmap27init_mmio_platform_resource>:
   1407c:       e92d4bf0        push    {r4, r5, r6, r7, r8, r9, fp, lr}
   14080:       e24dd0a8        sub     sp, sp, #168    ; 0xa8
   14084:       e1a04000        mov     r4, r0
   14088:       e5900000        ldr     r0, [r0]
   1408c:       e1a06002        mov     r6, r2
   14090:       ebfffffe        bl      0 <devm_platform_ioremap_resource>
   14094:       e1a05000        mov     r5, r0
   14098:       ebfffffe        bl      0 <rust_helper_is_err>
   1409c:       e3500000        cmp     r0, #0
   140a0:       1a000029        bne     1414c <_RNvMs0_NtCsbDqzXfLQacH_6kernel6regmapNtB5_6Regmap27init_mmio_platform_resource+0xd0>
```

Reply:
Note that I agree we need a template `ptr_err_check` or `ptr_to_result`, because the return type will use the template parameter, and that will empower us with some strong type checking. I also made `ptr_to_result` as a template function.

And if we think that `from_kernel_ptr` is worth discussing, I don't think the template parameter will buy us anything, it only appears at the input parameter, and the type information gets lost afterwards. So it's really unnecessary ;-(

Anyway thanks for looking into the generated code to prove I'm wrong ;-)

Reply:
> Anyway thanks for looking into the generated code to prove I'm wrong ;-)

I didn't want to imply that you were wrong - you were asking a question that I didn't know the answer to either (Would unnecessary template functions increase the code size?) and it sounded like an interesting question.

I apologize if it came across that way.

Reply:
> > Anyway thanks for looking into the generated code to prove I'm wrong ;-)
> 
> I didn't want to imply that you were wrong - you were asking a question that I didn't know the answer to either (Would unnecessary template functions increase the code size?) and it sounded like an interesting question.
> 
> I apologize if it came across that way.

No need to apologize ;-) Actually I'm glad you looked into it.

Note that I actually kinda answered my question myself with "For example, if we have from_kernel_ptr<i32> and from_kernel_ptr<i8>, there will be two functions instead of one.", so literally, I was wrong and you simply showed the answer is opposite.